source/common/upstream/cds_api_impl.cc

```diff
@@ -57,7 +57,16 @@ void CdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
     const std::string cluster_name = cluster.name();
     try {
       clusters_to_remove.erase(cluster_name);
-      if (cm_.addOrUpdateCluster(cluster, version_info)) {
+      if (cm_.addOrUpdateCluster(cluster, version_info, [this](auto, auto state) {
+            // Keep CDS paused as long as there is at least one cluster in the warming state.
+            if (state == ClusterManager::ClusterWarmingState::Starting &&
+                cm_.warmingClusterCount() == 1) {
+              cm_.adsMux().pause(Config::TypeUrl::get().Cluster);
```

Comment:
I'm worried here about 2 things:
1) Potential lifetime issues. Is there any chance that the API can go away during this operation? I don't think so, but just checking.
2) I'm more concerned about remove of a warming cluster. I don't think we will ever become unpaused. Is the idea here that removal can never happen while it's paused? Are we concerned about not having timeouts everywhere such this can potentially get fully stuck? 

I think some more comments here would be good around some of the implications of the pause, these concerns, and in general why we are doing the pause if you don't mind?

Reply:
I'll answer (2) now as I have studied this while I worked on the change. I'll follow up with (1) and provide code comments later.

While `ClusterManagerImpl` does allow removal of a warming cluster [here](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L542), `CdsApiImpl` is the components which drives the actual removal. `CdsApiImpl` can only removes active clusters because it initializes [`clusters_to_remove`](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cds_api_impl.cc#L55) variable from the list of [active clusters](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.h#L182-L186) it gets from the `ClusterManagerImpl`. So I think we can not end being stuck in the paused state as long as following two assumptions are true:

1. `ClusterManager::clusters()` should not return any cluster in the warming state.
2. No other component can pull the rug out from under `CdsApiImpl` by calling `ClusterManager::removeCluster()` on a warming cluster.

Do you think those assumptions are too strong to be made? If so, I think we could actually call the warming callbacks from the `ClusterManager::removeCluster()`, instead of introducing timeouts. This would require some kind of warming callback tracking per added cluster in the `ClusterManagerImpl`.

Reply:
Regarding (1). I think that's not possible. `CdsApiImpl` can be destructed before `ClusterManagerImpl`, but this happens after the main dispatch loop exited, so I assume warming callback would not be called at that time.

Reply:
> Do you think those assumptions are too strong to be made? If so, I think we could actually call the warming callbacks from the `ClusterManager::removeCluster()`, instead of introducing timeouts. This would require some kind of warming callback tracking per added cluster in the `ClusterManagerImpl`.

Yeah, storing the warming callback and calling it during removal is what I had in mind, but as long as you don't think it can happen (I agree), I think it's fine to leave it as is with good comments to warn us if things change in the future.

> Regarding (1). I think that's not possible. `CdsApiImpl` can be destructed before `ClusterManagerImpl`, but this happens after the main dispatch loop exited, so I assume warming callback would not be called at that time.

Agreed, let's just cover with a good comment.